README.TXT

** SEE TEST MODEL NOTES AT BOTTOM OF THIS DOCUMENT **


How to use Display Stream Compression (DSC) algorithm model
--------------------------------------------------------------

1)  The typical usage for the DSC model is:

DSC -F <configuration_file>

The configuration_file contains a list of options for the model.  An example is given in test.cfg that is included with the model.  The parameters are listed in the section below.


2)  The configuration file can specify whether the DSC executable behaves as an encoder, decoder, or both using the FUNCTION parameter.  If the FUNCTION is 1, the executable will encode to an output bitstream that is stored in a file (or files) with the same base name and a .dsc extension.  If the FUNCTION is set to 2, the executable will decode a .dsc file (or files) specified.  If .dpx or .ppm files are listed in the input list file instead of .dsc files, the .dsc files will be decoded and PSNR will be calculated as well.  If the FUNCTION is 0, the executable will encode and decode, but no bitstream file will be generated.

A list file (SRC_LIST) is required to indicate which source files to encode.  The list file is also required in decode mode to indicate which .dsc files to decode.  The list file may contain full pathnames.  By default, the model outputs to the current working directory; however, an alternative OUT_DIR can be specified in the config file.  If the same list file is used for encoding and decoding, the PSNR will be calculated between the source file and decoded output.

Parameters besides SRC_LIST and FUNCTION are ignored for the decoder, since the configuration is stored as part of the bitstream file.


3)  Output files to the current working directory:
	- log.txt (or filename specified as LOG_FILENAME) provides PSNR information
	- <name>.ref.dpx is the converted DPX file that represents the encoder input
	- <name>.out.dpx is the decoder output
        - <name>.dsc is the output bitstream if FUNCTION = 1


4)  The included EXE file is compiled under Win32, and some users may experience memory allocation errors for very large pictures.  Compiling the code using "x64" platform target under Windows or compiling under 64-bit Linux will fix these errors.



Configuration file paramters
-----------------------------

FUNCTION - 0=encode and decode (no bitstream file is produced), 1=encode only, 2=decode only

BITS_PER_PIXEL - The constant bit rate of the output stream (bits/pixel).  Can be specified as a floating point number.

BITS_PER_COMPONENT - The input DPX files will be converted to this number of bits per component if they aren't already in that format and coded using this value for bpc.

SIMPLE_422 - If set to 1, USE_YUV_INPUT must be set 1.  The input DPX files will be converted to YCbCr 4:2:2 (if they aren't already in that format) and coded that way.  If set to 0, the input DPX files will be converted and coded as 4:4:4.

VBR_ENABLE - Enable VBR mode.

USE_YUV_INPUT - If set 0, the input DPX files will be converted to RGB (if they aren't already in that format) and coded that way.  If set to 1, the input DPX files will be converted to YCbCr (if the aren't already in that format) and coded that way.

SLICE_WIDTH - If specified, sets the width of the rectangular slices.  If unspecified, slices are assumed to be the full picture width.

SLICE_HEIGHT - If specified, sets the height of the rectangular slices.  If unspecified, slices are assumed to be the full picture height.

SWAP_R_AND_B - Allows the swapping of the R & B components of the input files if the DPX file input has incorrect component ordering.

SWAP_R_AND_B_OUT - Allows the swapping of the R & B components of the output files if the DPX viewer has incorrect component ordering.

DPX_BUGS_OVERRIDE - If noncompliant DPX files are input, allows the DPX reader to read them properly.  This is required to be set = 2 for the Sharp images that were generated by ImageMagick.

DPX_READ_NOSWAP - Ignore magic number and don't byte-swap image data for DPX reading

DPX_WRITE_BSWAP - Write the DPX file endianness the opposite way from the host machine.

SRC_LIST - Specifies a text file that contains the DPX files to be processed (one file name per line).

INCLUDE - Several different include files are used for different rate control settings.  The distribution has recommended settings for 8, 10, and 12 bpc and 8 and 12 bpp.

BLOCK_PRED_ENABLE - Enables block prediction with a range of 8 positions.  A value of 0 disables block prediction.  Block prediction helps certain pathological images but can have a slight negative effect on "typical" content.

YUV_FILE_OUTPUT - Set to 1 to output .YUV files in 4:2:0 or 4:2:2 mode (output is only generated if it matches the coding pixel format)

DPX_FILE_OUTPUT (default 1) - Set to 0 to disable DPX file output

PPM_FILE_OUTPUT - Set to 1 to output PPM files

YUV_FILE_FORMAT - 0 = YUV files are planar 4:2:0; 1 = YUV files are packed (UYVY) 4:2:2




DPX Support Notes
------------------
The model supports reading and writing 8, 10, 12, and 16-bit DPX files.  We attempted to follow the SMPTE 268M specification as closely as possible.  Although some testing has been done to try to ensure compatibility, there may be some issues reading certain types of DPX files.  The written DPX files aren't guaranteed to work with any specific DPX reader software.

1)  See the included DPX application note for information on the DPX standard and compatibility as well as the config file options for DPX.

2)  We always output a .ref.dpx file that contains the image the codec is trying to code.  If your reader doesn't read this file, it won't be able to read the .out.dpx file either.  If you know your DPX reader ordinarily works with the .ref.dpx file, but some particular source image causes problems, there may be a compatibility issue between that source image and the DPX reader.

3)  The model can also input and output PPM format. Set PPM_FILE_OUTPUT to 1 to output PPM files.



The following features are added to DSC 1.2:
--------------------------------------------

-  Support for YUV 4:2:0 input/output .YUV files (only single frames currently supported).  Files with >8 BPC use 16-bit LE packing.
-  Support for native 420 and native 422 modes (set NATIVE_420 or NATIVE_422 equal to 1 in the config file).  Note that USE_YUV_INPUT is required to be 1 and SIMPLE_422 is required to be 0 for these modes.
-  Support for 14 & 16 bpc
-  Some minor modifications for flatness, rate control.

Brief description of parameters:

- SECOND_LINE_BPG_OFFSET - Specifies BPG offset for 2nd line (used in 4:2:0 mode)

- SECOND_LINE_OFFSET_ADJ - Specifies a portion of the initial offset to use for the 2nd line (used in 4:2:0 mode)

- DSC_VERSION_MINOR (default 2) - Set minor version to 1 to generate a DSC 1.1 compliant output.

- PIC_WIDTH, PIC_HEIGHT - Picture width & height for YUV input



The following features are added to DSC 1.2a:
---------------------------------------------

- FULL_ICH_ERR_PRECISION - 0 = truncate errors in encoder ICH decision, 1 = use full precision for errors in encoder ICH decision

- Note that Native 4:2:0 was not correctly supported in DSC 1.2 and was deprecated; the DSC 1.2a model provides a corrected implementation for Native 4:2:0.


CHANGELOG.TXT


Change log for Display Stream Compression C model
--------------------------------------------------

V1.57 / released 12/12/2016 (final DSC 1.2a)
--------------------------------------------

- Removed bits_per_pixel check in PPS since it created an accidental restriction in 16bpc Native 4:2:2 mode

- Added FULL_ICH_ERR_PRECISION to README.txt file

- Defaulted FULL_ICH_ERR_PRECISION to 0 in test.cfg. This change was suggested solely for consistency with DSC 1.2 and does not represent guidance on which setting to use.



V1.56 / released 11/1/2016 (proposed DSC 1.2a)
----------------------------------------------

- Renamed ALT_ENC_ICH_SELECTION to FULL_ICH_ERR_PRECISION. Added it to test.cfg.

- Adjusted automatic slice height division algorithm to round up rather than down to avoid problems with odd heights.



V1.55 / released 9/23/2016
--------------------------

- Added SECOND_LINE_BPG_OFFSET and SECOND_LINE_OFFSET_ADJ to 4:2:0 RC cfg files.

- Fixed computation of default FIRST_LINE_BPG_OFFSET to use correct value of uncompressed_bpg

- Fixed computation of default SECOND_LINE_BPG_OFFSET to use nominally 12 (was 512, which was a typo/invalid)



V1.54 / released 9/7/2016
-------------------------

- Fixed bug in ICH decision



V1.53 / released 9/2/2016
-------------------------

- Added missing SCR that fixes 10/12-bit UYVY file read



V1.52 / released 9/1/2016
-------------------------

- ALT_ENC_ICH_SELECTION now only consists of using the extra precision for the errors.

- Clamp max QP in bit save mode so it does not exceed the valid range of QP values

- Reverted 4:2:0 PPS files to 1.48 versions.

- Fixed rcb_bits calculation for decoder.

- Changed auto slice height calculation to ensure even slice height in 4:2:0 mode.

- Fixed "misleading indent" warnings in Linux.



V1.51 / released 8/15/2016
--------------------------

- Reverted PPS files for 4:2:2 to 1.48 versions. Included 4:2:0 versions in orig_420_cfg subdirectory.

- Added default value for SECOND_LINE_BPG_OFFSET. Capped FIRST_LINE_BPG_OFFSET default so first line bpg rate does not exceed uncompressed bpg rate.

- Added warning message for RC_MAX_QP[14] that is too low and might cause overflow.

- Fixed bug in DPX output colorspace when running with FUNCTION==2.

- Removed unused parameter from GetQpAdjPredSize(). Fixed compiler warnings in Linux.

- Modified ALT_ENC_ICH_SELECTION:
  - Added ALT_ENC_ICH_SELECTION == 2 which is the same as ALT_ENC_ICH_SELECTION == 1 except using the 1.48 ICH decision logic.
  - Dropped ICH search reordering



V1.50 / released 6/28/2016
--------------------------

- Clarified error message for RGB mismatched with 4:2:2/4:2:0 mode

- In 4:2:0 mode, applied second_line_ofs_adj to rcXformOffset (state variable) instead of initial_offset (PPS variable).

- Fixed possible seg fault in certain modes when running with FUNCTION = 2

- Adjusted ALT_ENC_ICH_SELECTION to be more conservative. (consider dropping aspects other than higher error precision?)

- Modified max_qp for bit save mode in rate control so that it doesn't exceed the max QP for range 14



V1.49 / released 5/25/2016
--------------------------

- Added ALT_ENC_ICH_SELECTION mode for reducing subtle artifacts in some cases in flat areas

- Fixed nsl_bpg_offset bounds check

- Added QP bounds check for DSC 1.1 YCbCr mode to avoid qLevel == cpntBitDepth

- Fixed chunk_sizes[] malloc to use sizeof(int) rather than sizeof(int *)

- For native 4:2:0, ICH no longer references previous line's samples in second line

- Fixed condition for ICH reset on slice lines to use correct value for slice width

- Fixed BP search in native 4:2:0 mode

- Fixed computation of predActivity in rate control for native 4:2:0 mode

- Fixed debug printout for decoding ICH indices

- For native 4:2:0, use LEFT instead of MAP prediction for first and second lines for chroma

- Ensure 16bpc YCoCg conversion does not exceed 16 bits

- Fatal error if slice width or slice height is not an even multiple of 2

- Revised config files for improved readability and 4:2:2/4:2:0 improved PPS values



V1.48 / released 12/14/2015 (DSC 1.2 final)
---------------------------

- Removed a reserved byte from PPS so some long DSC 1.2 fields would be line up with 2-byte boundaries

- Removed old, unused code from GetQpAdjPredSize()

- Added multi-frame YUV read/write

- Added independent range checks for DSC 1.1 and 1.2 modes for BITS_PER_COMPONENT and LINEBUF_DEPTH

- Fix for planar YUV read (12 bit and higher)



V1.47 / released 10/21/2015 (TEST MODEL FOR 4:2:2)
---------------------------

- Fixed ICH for native modes so the pixels from the previous line are correctly centered over the current group

- Adjusted order of ICH indices in native 4:2:2 mode to keep ICH indices consistent for chroma regardless of mode



V1.46 / released 10/16/2015 (TEST MODEL FOR 4:2:2)
---------------------------

- Changed native 4:2:2 mode to be similar to native 4:2:0 except with 4 components (even and odd luma, Cb, and Cr).

- Fixed small memory leak in codec_main.c

- Fixed PopulateOrigLine() to ensure out-of-slice pixels aren't accidentally used



V1.45 / released 10/13/2015
---------------------------

- Removed flatness modification (except changes related to first line), ICH modifiedQp adjustment, and MPP modification.

- Minor adjustments to RC to make bit saving mode slightly more conservative and allow the QP to drop to minQp-4 when groups are coded with no residuals.

- Minor adjustment to ICH decision to use DSC 1.1 mode when next group is flat. Fix calculation for 4:2:2 ICH decision.

- Changed rbsMin calculation to only use new calculation method for native 4:2:2 and native 4:2:0.

- Add flatness check 3 for native 4:2:2 mode

- Bug fix: somewhat_flat_qp_thresh and very_flat_qp not set correctly in codec_main.c



V1.44 / released 9/22/2015
--------------------------

- Updated PPS guidance to match MIPI recommendations


dsc_codec.py
# DSC_CODEC Python Class
# THIS CLASS IS IMPLEMENTED TO CONVERT IMAGE FILE TO PPM EXTENSION IMAGE
# CONVERTED FILE WILL BE USED TO COMPRESS AND DECODE USING DSC C MODEL
# CODED BY IM SEUNGWOO

#          DRAFT : 2020.05.14
# FINAL REVISION : 2020.05.18

import os
from PIL import Image
import time
import shutil
from init_pps import *

class DSC_CODEC:

    def __init__(self):
        self.is_cfg_created = False
        self.is_image_converted = False
        self.is_list_created = False
        self.print_eps = True

    def dsc_set_cfg(self, cfg_dict = None, cfg_file_name = "test_py.cfg"):
        self.cfg_dict = cfg_dict
        self.cfg_file_name = cfg_file_name
        self.cfg_line = {}

        if cfg_dict["Bits_per_component"] not in [8, 10, 12, 14, 16]:
            raise ValueError("%d is unsupported 'Bits_per_component Value'" %cfg_dict["Bits_per_component"])

        with open("test.cfg", 'r') as cfg:
            for num, lines in enumerate(cfg):
                self.cfg_line[num] = lines

        self.cfg_line[6] = self.cfg_line[6][:10] + str(self.cfg_dict["DSC_MODE"]) + self.cfg_line[6][11:]
        self.cfg_line[31] = self.cfg_line[31][:14] + str(self.cfg_dict["slice_width"]) #+ self.cfg_line[31][15:]
        self.cfg_line[34] = self.cfg_line[34][:15] + str(self.cfg_dict["slice_height"])# + self.cfg_line[34][16:]
        self.cfg_line[38] = self.cfg_line[38][:21] + str(self.cfg_dict["BLOCK_PRED_ENABLE"]) + self.cfg_line[38][22:]
        self.cfg_line[43] = self.cfg_line[43][:15] + str(self.cfg_dict["USE_YUV_INPUT"]) + self.cfg_line[43][16:]
        self.cfg_line[44] = self.cfg_line[44][:15] + str(self.cfg_dict["SIMPLE_422"]) + self.cfg_line[44][16:]
        self.cfg_line[45] = self.cfg_line[45][:15] + str(self.cfg_dict["NATIVE_422"]) + self.cfg_line[45][16:]
        self.cfg_line[46] = self.cfg_line[46][:15] + str(self.cfg_dict["NATIVE_420"]) + self.cfg_line[46][16:]
        self.cfg_line[48] = self.cfg_line[48][:23] + str(self.cfg_dict["FULL_ICH_ERR_PRECISION"]) + self.cfg_line[48][24:]
        self.Bits_per_component = self.cfg_dict["Bits_per_component"]
        self.Bits_per_pixel = self.cfg_dict["Bits_per_pixel"]

        if self.cfg_dict["NATIVE_422"]:
            self.cfg_line[61] = self.cfg_line[61][:11] + "rc_%dbpc_%dbpp_422.cfg\n" % (self.Bits_per_component, self.Bits_per_pixel)
            self.cfg_line[84] = self.cfg_line[84][:16] + " 1" + self.cfg_line[84][17:]
        elif self.cfg_dict["NATIVE_420"]:
            self.cfg_line[61] = self.cfg_line[61][:11] + "rc_%dbpc_%dbpp_420.cfg\n" % (self.Bits_per_component, self.Bits_per_pixel)
            self.cfg_line[84] = self.cfg_line[84][:16] + " 0" + self.cfg_line[84][17:]
        else:
            self.cfg_line[61] = self.cfg_line[61][:11] + "rc_%dbpc_%dbpp.cfg\n" % (self.Bits_per_component, self.Bits_per_pixel)
            self.cfg_line[84] = self.cfg_line[84][:16] + " 0" + self.cfg_line[84][17:]

        try:
            with open(self.cfg_file_name, "wt") as cfg_py:
                for txt in self.cfg_line.values():
                    cfg_py.write(txt)
            print("Setting " + '"' + self.cfg_file_name + '"' + " file...")
            self.is_cfg_created = True

            return self

        except:
            raise NotImplementedError

    def convert_input_images(self, image_path = 'input_files/', output_extension = "ppm"):
        if self.cfg_dict["DSC_MODE"] == 2:
            self.is_image_converted = True
            print("Skip Image Converting Process in Decode Mode")
            return self

        self.image_path = image_path
        self.output_extension = output_extension
        self.input_file_list = os.listdir(self.image_path)
        self.output_file_list = []
        print("Converting Input Images to " + '"' + self.output_extension.upper() + '"' + " Extension")

        try:
            for fname in self.input_file_list:
                self.output_extension = output_extension
                name_index = fname.find(".")
                start_t = time.time()
                file_fmt = fname[-4:]

                if (file_fmt == ".dpx") or (file_fmt == ".ppm") or (file_fmt == ".yuv"):
                    print("This %s file is no need to be converted, just use this file" %fname)
                    self.output_extension = fname[-4:]
                    self.print_eps = False
                    shutil.copy(self.image_path + fname, os.getcwd())
                    fname_ = fname[: name_index] + self.output_extension

                else:
                    fname_ = fname[: name_index] + self.output_extension
                    im = Image.open(image_path + fname)
                    im.save(os.getcwd() + "/" + fname_)

                end_t = time.time()
                if self.print_eps:
                    print("Elapsed Time for convert Current Image : %.3fs" %(end_t - start_t))

                #fname_ = fname[: name_index] + self.output_extension
                self.output_file_list.append(fname_)
                self.is_image_converted = True

            return self

        except:
            raise NotImplementedError

    def dsc_set_image_list(self):
        if self.cfg_dict["DSC_MODE"] == 2:
            self.is_list_created = True
            print("Skip Setting Image List Process in Decode Mode")
            return self

        try:
            with open("test_list.txt", "w") as f:
                for f_name in self.output_file_list:
                    f.write(f_name + "\n")
            print("Writing Input File List")
            self.is_list_created = True

            return self

        except:
            raise NotImplementedError

    def start_dsc(self):
        self.is_dsc_ready = self.is_cfg_created & self.is_image_converted & self.is_list_created

        if self.is_dsc_ready:
            print("Starting DSC Model")
            self.dsc_path = os.getcwd()
            os.system("cd {}".format(self.dsc_path))
            os.system("DSC -F " + self.cfg_file_name)

            return self

        else:
            raise NotImplementedError("DSC is not ready to start")